network/endpoint/endpoint.go

```diff
@@ -4,62 +4,79 @@ import (
 	"context"
 	"time"
 
-	"github.com/libp2p/go-libp2p/core/network"
 	"github.com/plprobelab/go-kademlia/kad"
 	"github.com/plprobelab/go-kademlia/network/address"
-	"github.com/plprobelab/go-kademlia/network/message"
+)
+
+// Connectedness signals the capacity for a connection with a given node.
+// It is used to signal to services and other peers whether a node is reachable.
+type Connectedness int
+
+const (
+	// NotConnected means no connection to peer, and no extra information (default)
+	NotConnected Connectedness = iota
+
+	// Connected means has an open, live connection to peer
+	Connected
+
+	// CanConnect means recently connected to peer, terminated gracefully
+	CanConnect
+
+	// CannotConnect means recently attempted connecting but failed to connect.
+	// (should signal "made effort, failed")
+	CannotConnect
 )
 
 // RequestHandlerFn defines a function that handles a request from a remote peer
-type RequestHandlerFn[K kad.Key[K]] func(context.Context, address.NodeID[K],
-	message.MinKadMessage) (message.MinKadMessage, error)
+type RequestHandlerFn[K kad.Key[K]] func(context.Context, kad.NodeID[K],
+	kad.Message) (kad.Message, error)
 
 // ResponseHandlerFn defines a function that deals with the response to a
 // request previously sent to a remote peer.
-type ResponseHandlerFn[K kad.Key[K]] func(context.Context, message.MinKadResponseMessage[K], error)
+type ResponseHandlerFn[K kad.Key[K], A kad.Address[A]] func(context.Context, kad.Response[K, A], error)
 
 // Endpoint defines how Kademlia nodes interacts with each other.
-type Endpoint[K kad.Key[K]] interface {
+type Endpoint[K kad.Key[K], A kad.Address[A]] interface {
 	// MaybeAddToPeerstore adds the given address to the peerstore if it is
 	// valid and if it is not already there.
-	MaybeAddToPeerstore(context.Context, address.NodeAddr[K], time.Duration) error
+	MaybeAddToPeerstore(context.Context, kad.NodeInfo[K, A], time.Duration) error
```

Comment:
I wonder if `MaybeAddToPeerstore` and `NetworkAddress` actually belong on `NetworkedEndpoint`

Reply:
I'm not as deep into the Endpoint abstraction to really judge :/